art_tree_test.go

```diff
@@ -729,8 +729,111 @@ func TestInsertWithSameByteSliceAddress(t *testing.T) {
 
 	for k, _ := range keys {
 		n := tree.Search([]byte(k))
-		if n == nil{
+		if n == nil {
 			t.Errorf("Did not find entry for key: %v\n", []byte(k))
 		}
 	}
 }
+
+func TestPrefixSearch(t *testing.T) {
+	tree := NewArtTree()
+
+	searchWords := []string{
+		"abcd", "abde", "abfg", "abgh",
+		"abcfgh", "abezyx",
+		"bcdef", "bcdi", "bcdgh", "abef",
+	}
+
+	for _, s := range searchWords {
+		tree.Insert([]byte(s), s)
+	}
+	res := tree.Search([]byte("abezyx"))
+	if res != "abezyx" {
+		t.Error("Unexpected search result.")
+	}
+
+	rr := tree.PrefixSearch([]byte("x"))
+	if rr == nil {
+		t.Error("empty results empty arrays")
+	} else if len(rr) > 0 {
+		t.Error("shouldn't be found", rr)
+	}
+
+	rr = tree.PrefixSearch([]byte("ax"))
+	if rr == nil {
+		t.Error("empty results empty arrays")
+	} else if len(rr) > 0 {
+		t.Error("shouldn't be found", rr)
+	}
+
+	rs := ""
+	for res := range tree.PrefixSearchChan([]byte("abc")) {
+		rs += res.Value.(string) + ","
```

Comment:
all of the test cases process the prefix searches as a series of results.  but all of the other ones (i only modified this one to use the channel), you can see we made an array of results somewhere, but only used it to scan.  this channel case is more interesting, because we are receiving the results and processing them inline.